### ISSUE TITLE
Point Image Serialization - Shapefiles

### ISSUE THREAD

**Issue by ultraTCS** [MapWindow/MapWinGIS#47]
There appears to be a bug in image serialization - or deserialization. 

I can apply an image to a point shape. I am using a png, so it is not serializing as GdiPlus. Put/Get Image API works fine. However after I save the layer and then re-load it, 3/4 of the image is black. Perhaps this has to do with the alpha channel? Perhaps with base64 encoding/decoding?

TIA,
LT

**Comment by ultraTCS** [comment]
Can someone please confirm this behavior?
Thanks,
LT

**Comment by pmeems** [comment]
Can you add the code you are using so we can easily use it as a unit test?
Then we can confirm or not this behavior.
Here you can find an example of such a unit test:
https://github.com/MapWindow/MapWinGIS/blob/master/unittests/MapWinGISTests/ImageTests.cs

**Comment by ultraLT** [comment]
No longer an issue and unrelated to MapWinGIS. Closing.

**Comment by ultraTCS** [comment]
Reopening this bug as I can give you a simple example to recreate. I'm using C++ so can't add to unit tests.
Bug occurs only for an image created in memory as opposed to opened from disk.
```
// open a point type shapefile
// ..
// create image object - 2x2 red square
int size = 2;
VARIANT_BOOL res = mapImage->CreateNew(size, size);
for (int i = 0; i < size; ++i)
   for (int j = 0; j < size; ++j)
      mapImage->PutValue(i, j, RGB(255, 0, 0));

shapeFile->DefaultDrawingOptions->PutPointType(MapWinGIS::ptSymbolPicture);
shapeFile->DefaultDrawingOptions->PutPicture(mapImage);
mapImage->Save("C:\\temp\\bmp1.png", VARIANT_FALSE, ImageType::USE_FILE_EXTENSION);   // save before serializing
map->SaveLayerOptions(layerId, "", VARIANT_TRUE, "");
map->LoadLayerOptions(layerId, "");
mapImage = shapeFile->DefaultDrawingOptions->GetPicture();
mapImage->Save("C:\\temp\\bmp2.png", VARIANT_FALSE, ImageType::USE_FILE_EXTENSION);


```
When I execute this one of the pixels in bmp2 is black.
TIA

**Comment by ultraTCS** [comment]
@pmeems 
I've found the issue here.
When image is encoded/decoded, the buffer length is too small; it doesn't consider the alpha channel in the image data (colour object).
So line 3043 in Image.cpp should be:
`std::string s = base64_encode(data, bufferWidth * bufferHeight * 4);`
Similarly, line 3112:
`memcpy(_imageData, data, sizeof(unsigned char) * width * height * 4);`

**Comment by ultraTCS** [comment]
Also I believe another bug in same file....line 3108 should read:
`_sourceType = istInMemory;`

**Comment by jerryfaust** [comment]
@ultraLT  Thank you for working through this.  It is helpful.  I'll step through to confirm your changes, and should be able to get them into the release.

**Comment by ultraTCS** [comment]
thanks @jerryfaust 

fix is in v4.9.5: https://mapwindow.atlassian.net/browse/MWGIS-97
